csharp/ql/lib/semmle/code/csharp/dataflow/internal/ContentDataFlow.qll

```diff
@@ -1,185 +1,215 @@
-     *    Assume that sinks have some existing call context to disallow
-     *    conflicting argument-to-parameter flow directly preceding the sink.
-     * - `FeatureEqualSourceSinkCallContext`:
-     *    Implies both of the above and additionally ensures that the entire flow
-     *    path preserves the call context.
-     */
-    FlowFeature getAFeature() { none() }
+    predicate isAdditionalFlowStep(
+      DataFlow::Node node1, FlowState state1, DataFlow::Node node2, FlowState state2
+    ) {
+      storeStep(node1, state1, _, node2, state2) or
+      readStep(node1, state1, _, node2, state2) or
+      additionalStep(node1, state1, node2, state2)
+    }
 
-    /** Gets a limit on the number of reads out of sources and number of stores into sinks. */
-    int accessPathLimit() { result = DataFlowPrivate::accessPathLimit() }
+    predicate isAdditionalFlowStep = ContentConfig::isAdditionalFlowStep/2;
 
-    /** Holds if `c` is relevant for reads out of sources or stores into sinks. */
-    predicate isRelevantContent(ContentSet c) { any() }
+    predicate isBarrier = ContentConfig::isBarrier/1;
 
-    /**
-     * Holds if data stored inside `sourceAp` on `source` flows to `sinkAp` inside `sink`
-     * for this configuration. `preservesValue` indicates whether any of the additional
-     * flow steps defined by `isAdditionalFlowStep` are needed.
-     *
-     * For the source access path, `sourceAp`, the top of the stack represents the content
-     * that was last read from. That is, if `sourceAp` is `Field1.Field2` (with `Field1`
-     * being the top of the stack), then there is flow from `source.Field2.Field1`.
-     *
-     * For the sink access path, `sinkAp`, the top of the stack represents the content
-     * that was last stored into. That is, if `sinkAp` is `Field1.Field2` (with `Field1`
-     * being the top of the stack), then there is flow into `sink.Field1.Field2`.
-     */
-    final predicate hasFlow(
-      Node source, AccessPath sourceAp, Node sink, AccessPath sinkAp, boolean preservesValue
-    ) {
-      exists(DF::PathNode pathSource, DF::PathNode pathSink |
-        this.(ConfigurationAdapter).hasFlowPath(pathSource, pathSink) and
-        nodeReaches(pathSource, TAccessPathNil(), TAccessPathNil(), pathSink, sourceAp, sinkAp) and
-        source = pathSource.getNode() and
-        sink = pathSink.getNode()
-      |
-        pathSink.getState().(InitState).decode(preservesValue)
-        or
-        pathSink.getState().(ReadState).decode(_, preservesValue)
-        or
-        pathSink.getState().(StoreState).decode(_, preservesValue)
-      )
+    predicate isBarrier(DataFlow::Node node, FlowState state) { none() }
+
+    DataFlow::FlowFeature getAFeature() { result = ContentConfig::getAFeature() }
+
+    // needed to record reads/stores inside summarized callables
+    predicate includeHiddenNodes() { any() }
+  }
+
+  private module Flow = DataFlow::GlobalWithState<FlowConfig>;
+
+  import Flow
```

Comment:
I think this should be private.

Reply:
We don't need the import at all if we only expose the specialzied `flow` predicate.